library/alloc/src/boxed.rs

```diff
@@ -1776,55 +1782,16 @@ impl<T: Clone, A: Allocator + Clone> Clone for Box<T, A> {
     /// ```
     #[inline]
     fn clone_from(&mut self, source: &Self) {
-        (**self).clone_from(&(**source));
-    }
-}
-
-#[cfg(not(no_global_oom_handling))]
-#[stable(feature = "box_slice_clone", since = "1.3.0")]
-impl<T: Clone, A: Allocator + Clone> Clone for Box<[T], A> {
-    fn clone(&self) -> Self {
-        let alloc = Box::allocator(self).clone();
-        self.to_vec_in(alloc).into_boxed_slice()
-    }
-
-    /// Copies `source`'s contents into `self` without creating a new allocation,
-    /// so long as the two are of the same length.
-    ///
-    /// # Examples
-    ///
-    /// ```
-    /// let x = Box::new([5, 6, 7]);
-    /// let mut y = Box::new([8, 9, 10]);
-    /// let yp: *const [i32] = &*y;
-    ///
-    /// y.clone_from(&x);
-    ///
-    /// // The value is the same
-    /// assert_eq!(x, y);
-    ///
-    /// // And no allocation occurred
-    /// assert_eq!(yp, &*y);
-    /// ```
-    fn clone_from(&mut self, source: &Self) {
-        if self.len() == source.len() {
-            self.clone_from_slice(&source);
+        // The following comparison may fail for vtable pointers even if both
+        // types are the same, in which case we take the slow path.
+        if ptr::metadata(self) == ptr::metadata(source) {
```

Comment:
```suggestion
        if ptr::metadata::<T>(self) == ptr::metadata::<T>(source) {
```

Reply:
This change should fix the [clippy lint firing](https://triage.rust-lang.org/gha-logs/rust-lang/rust/50600845571) (and is necessary for correctness, as otherwise this compares the metadata of `&Box<T>` which is always equal as `&Box<T>` is a thin pointer).